src/inc/corjitflags.h

```diff
@@ -143,7 +143,7 @@ class CORJIT_FLAGS
         CORJIT_FLAG_USE_LZCNT               = 51,
         CORJIT_FLAG_USE_PCLMULQDQ           = 52,
         CORJIT_FLAG_USE_POPCNT              = 53,
-        CORJIT_FLAG_UNUSED23                = 54,
+        CORJIT_FLAG_USE_BASE                = 54,
```

Comment:
It unlocks `opts.setSupportedISA(InstructionSet_Base);`. I haven't actually investigated what that does, but since it's in the giant if block, I assumed we need to guard it somehow.

Reply:
`InstructionSet_Base` means instructions available on all platforms. It is not used for x64 currently. 

ARM64 uses for methods that are available on every ARM64 hardware: https://github.com/dotnet/coreclr/blob/master/src/System.Private.CoreLib/shared/System/Runtime/Intrinsics/Arm/Arm64/Base.cs

Reply:
> InstructionSet_Base means instructions available on all platforms. It is not used for x64 currently.

SSE and SSE2 are in that boat on x64 though (couple lines below the `opts.setSupportedISA(InstructionSet_Base);`). I think we still need a new flag to cover those cases, no? I could name it differently if there's a better name, but BASE matches what it means for ARM64.

Reply:
You are right that SSE and SSE2 are always available on the currently supported Intel system. SSE and SSE2 have their own API class (and everything else that comes with it) for consistency over time. It was an explicit hardware intrinsic decision. E.g. AVX may be always available on the currently supported Intel systems, but they will have their own separate API class for AVX.

Base is different. Base are instructions that have been always available on the given platform. `InstructionSet_Base` enum value exists as an internal implementation detail in the JIT. I do not think it makes sense to have it in the ` CORJIT_FLAG_USE_` bit field on the JIT/EE interface. It would be always set, for all JIT, R2R and full AOT.

Reply:
@tannergooding @fiigii @CarolEidt Could you please chime in on whether you think we need the `BASE` flag on JIT/EE interface?

Reply:
I believe there should be a JIT/EE configuration knob that allows disabling the HWIntrinsic support altogether. For CoreCLR, we effectively have that already via the `COMPlus_EnableHWIntrinsic` configuration knob (which we determined was better than `EnableBase` or some variant on it).

We currently are exposing the base `EnableHWIntrinsic` flag and one flag per ISA exposed. The flags are hierarchical in nature, so setting `EnableHWintrinsic=0` will disable all HWIntrinsic support, while setting `EnableSSE=0` would disable the SSE HWIntrinsics and any ISAs dependent on that (so SSE2, AES, PCLMULQDQ, SSE3, SSSE3, SSE4.1, SSE4.2, POPCNT, AVX, FMA, and AVX2, but not BMI1, BMI2, or LZCNT). The flags are exposed so that developers can test and validate their current hardware and any downlevel code-paths, without needing to recompile.

I believe the same system would also be beneficial to other targets (such as CoreRT or AOT tagets in general). That is, the user should be able to opt-out of HWIntrinsics entirely or they should be able to opt-into higher level ISAs where possible (the default behavior should likely be the same as CoreCLR, which is SSE/SSE2 as a baseline with HWIntrinsics enabled).

Reply:
Just as a note, although it might be obvious: Setting `EnableSSE=0` only disables the HWIntrinsic support, it does not disable the SIMD Vector support (the types under `System.Numerics.Vectors` such as `Vector2/3/4` and `Vector<T>`) nor does it disable the regular `SSE/SSE2` codegen. However, some flags (like `EnableAVX`) do have a slightly broader impact and impact both the HWIntrinsic and SIMD Vector codegen. In some cases it also impacts general codegen or some general purpose optimizations that are done (because it, not being a baseline ISA, actually gets disabled in the compiler).

Reply:
> For CoreCLR, we effectively have that already via the COMPlus_EnableHWIntrinsic

For CoreRT compiler (and other AOT targets in general), we have it as well. You can specify any of these custom JIT config switches and overrides on the compiler command line.

Given this, do you believe that we need `JIT_FLAG_...` flag to enable/disable all hardware intrinsics?

Reply:
Let me double check what you want me to do here, because it still doesn't make sense to me.

You want SSE/SSE2 to be fully controlled by a ComPlus environment variable ("JitConfig.EnableHWIntrinsic"in combination with "JitConfig.EnableSSE"). The default for these values is [enabled](https://github.com/dotnet/coreclr/blob/5c9febf6346f90c06e70b3cb9e002300e50fa54b/src/jit/jitconfigvalues.h#L237), so I'll need to find a place to overwrite the default for EnableHWIntrinsic with a new default somewhere on the EE side (where?) when JIT is triggered from Crossgen to keep the existing behavior. This will also allow the user to specify EnableHWIntrinsic using an environment variable before launching crossgen to enable code paths that were not allowed before.

SSE/SSE2 will be different from how we control other ISAs that use [a combination](https://github.com/dotnet/coreclr/blob/5c9febf6346f90c06e70b3cb9e002300e50fa54b/src/jit/compiler.cpp#L2388) of a JIT flag (master enable switch), and the environment value (that only allows opting out).

Reply:
I think all you need to do is to delete the `jitFlags.IsSet(JitFlags::JIT_FLAG_PREJIT)` check in the JIT. It is redundant.

crossgen suppresses generation of any hardware intrinsics code today via an early check in the type loader. Look for `COMPlusThrow(kTypeLoadException, IDS_EE_HWINTRINSIC_NGEN_DISALLOWED)`. Removing the `jitFlags.IsSet(JitFlags::JIT_FLAG_PREJIT)` check should have no affect on crossgen generate code.

Reply:
I though the setSupportedISA calls allow more than just recognition of intrinsics (e.g. they would allow codegen to use those special instructions when it deems appropriate in normal code). This makes things simpler.